### PR TITLE
Fix enrollment API non returning data for Data Element linked to Stage

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
- DHIS2-6931
- Minor refactoring of JdbcEventAnalyticsManager and JdbcEnrollmentAnalyticsManager
- Added unit tests

### CODE REVIEW

**Comment by jason-p-pickering** [comment]
Can one of the admins verify this patch?

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 1]
```diff
@@ -110,7 +111,7 @@ public AbstractJdbcEventAnalyticsManager( @Qualifier( "readOnlyJdbcTemplate" ) J
      *
      * @param params the {@link EventQueryParams}.
      */
-    protected String getPagingClause( EventQueryParams params, int maxLimit )
+    private String getPagingClause(EventQueryParams params, int maxLimit)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]:
```suggestion
    private String getPagingClause( EventQueryParams params, int maxLimit )
```

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 2]
```diff
@@ -131,7 +132,7 @@ else if ( maxLimit > 0 )
      *
      * @param params the {@link EventQueryParams}.
      */
-    protected String getSortClause( EventQueryParams params )
+    private String getSortClause(EventQueryParams params)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]:
```suggestion
    private String getSortClause( EventQueryParams params )
```

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 3]
```diff
@@ -161,7 +162,7 @@ protected String getSortClause( EventQueryParams params )
      * When grouping with non-default analytics period boundaries, all periods are skipped in the group
      * clause, as non default boundaries is defining their own period groups within their where clause.
      */
-    protected List<String> getGroupByColumnNames( EventQueryParams params )
+    private List<String> getGroupByColumnNames(EventQueryParams params)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]:
```suggestion
    private List<String> getGroupByColumnNames( EventQueryParams params )
```

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 4]
```diff
@@ -172,7 +173,7 @@ protected String getSortClause( EventQueryParams params )
      * In the case of non-default boundaries{@link EventQueryParams#hasNonDefaultBoundaries},
      * the period is hard coded into the select statement with "(isoPeriod) as (periodType)".
      */
-    protected List<String> getSelectColumns( EventQueryParams params )
+    protected List<String> getSelectColumns(EventQueryParams params)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]:
```suggestion
    protected List<String> getSelectColumns( EventQueryParams params )
```

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 5]
```diff
@@ -552,7 +561,7 @@ protected String getOrgUnitAlias( EventQueryParams params )
      * @param params the event query parameters.
      * @param dimensionType the dimension type.
      */
-    protected String getAlias( EventQueryParams params, DimensionType dimensionType )
+    private String getAlias(EventQueryParams params, DimensionType dimensionType)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]:
```suggestion
    private String getAlias( EventQueryParams params, DimensionType dimensionType )
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+            + " as ax where enrollmentdate >= '2017-01-01' and enrollmentdate <= '2017-12-31' and (uidlevel0 = 'ouabcdefghA' ) limit 10001";
+
+        assertSql( expected, sql.getValue() );
+
+    }
+
+    @Test
+    public void verifyWithProgramStageAndNumericDataElement()
+    {
+        verifyWithProgramStageAndNumericDataElement(ValueType.NUMBER);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
        verifyWithProgramStageAndNumericDataElement( ValueType.NUMBER );
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+    @Test
+    public void verifyWithProgramStageAndTextDataElement()
+    {
+        verifyWithProgramStageAndNumericDataElement( ValueType.TEXT );
+    }
+
+    private  void verifyWithProgramStageAndNumericDataElement(ValueType valueType) {
+
+        EventQueryParams params = createRequestParams(this.programStage, valueType);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
        EventQueryParams params = createRequestParams( this.programStage, valueType );
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        verifyWithProgramStageAndNumericDataElement(ValueType.NUMBER);
+    }
+
+    @Test
+    public void verifyWithProgramStageAndTextDataElement()
+    {
+        verifyWithProgramStageAndNumericDataElement( ValueType.TEXT );
+    }
+
+    private  void verifyWithProgramStageAndNumericDataElement(ValueType valueType) {
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
    private  void verifyWithProgramStageAndNumericDataElement( ValueType valueType ) {
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+            + " as ax where ax.\"monthly\" in ('2000Q1') and (uidlevel0 = 'ouabcdefghA' ) "
+            + "and ps = '" + programStage.getUid() + "' limit 101";
+
+        assertSql( expected, sql.getValue()   );
+    }
+
+    @Test
+    public void verifyWithProgramStageAndTextualDataElementAndFilter() {
+
+        EventQueryParams params = createRequestParamsWithFilter(programStage, ValueType.TEXT);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
        EventQueryParams params = createRequestParamsWithFilter( programStage, ValueType.TEXT );
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+        subject.getEnrollments( params, new ListGrid(), 10000);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
        subject.getEnrollments( params, new ListGrid(), 10000 );
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+        String subSelect = "(select \"fWIAEtYVEGk\" from analytics_event_" + programA.getUid()  + " where analytics_event_"
+                + programA.getUid()  + ".pi = ax.pi and \"fWIAEtYVEGk\" is not null and ps = '"
+                + programStage.getUid() + "' order by executiondate desc limit 1 )";
+
+        String expected = "ax.\"monthly\",ax.\"ou\"," + subSelect + "  from " + getTable( programA.getUid() )
+                + " as ax where ax.\"monthly\" in ('2000Q1') and (uidlevel0 = 'ouabcdefghA' ) "
+                + "and ps = '" + programStage.getUid() + "' and lower(" + subSelect + ") > '10' limit 10001";
+
+        assertSql( expected, sql.getValue()   );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
        assertSql( expected, sql.getValue() );
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    @Test
+    public void verifyWithProgramStageAndNumericDataElementAndFilter2() {
+
+        EventQueryParams params = createRequestParamsWithFilter(programStage, ValueType.NUMBER);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
        EventQueryParams params = createRequestParamsWithFilter( programStage, ValueType.NUMBER );
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+        subject.getEnrollments( params, new ListGrid(), 10000);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
        subject.getEnrollments( params, new ListGrid(), 10000 );
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+        String subSelect = "(select \"fWIAEtYVEGk\" from analytics_event_" + programA.getUid()  + " where analytics_event_"
+                + programA.getUid()  + ".pi = ax.pi and \"fWIAEtYVEGk\" is not null and ps = '"
+                + programStage.getUid() + "' order by executiondate desc limit 1 )";
+
+        String expected = "ax.\"monthly\",ax.\"ou\"," + subSelect + "  from " + getTable( programA.getUid() )
+                + " as ax where ax.\"monthly\" in ('2000Q1') and (uidlevel0 = 'ouabcdefghA' ) "
+                + "and ps = '" + programStage.getUid() + "' and " + subSelect + " > '10' limit 10001";
+
+        assertSql( expected, sql.getValue()   );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EnrollmentAnalyticsManagerTest.java]:
```suggestion
        assertSql( expected, sql.getValue() );
```

**Pull request merged**